Content.Server/Fluids/EntitySystems/MoppingSystem.cs

```diff
@@ -0,0 +1,363 @@
+using Content.Server.Chemistry.Components.SolutionManager;
+using Content.Server.Chemistry.EntitySystems;
+using Content.Server.DoAfter;
+using Content.Server.Fluids.Components;
+using Content.Shared.Chemistry.Components;
+using Content.Shared.FixedPoint;
+using Content.Shared.Interaction;
+using Content.Shared.Popups;
+using Content.Shared.Sound;
+using Content.Shared.Tag;
+using Robust.Shared.Audio;
+using Robust.Shared.Player;
+using Robust.Shared.Map;
+using JetBrains.Annotations;
+
+namespace Content.Server.Fluids.EntitySystems;
+
+[UsedImplicitly]
+public sealed class MoppingSystem : EntitySystem
+{
+        [Dependency] private readonly DoAfterSystem _doAfterSystem = default!;
+        [Dependency] private readonly SpillableSystem _spillableSystem = default!;
+        [Dependency] private readonly TagSystem _tagSystem = default!;
+        [Dependency] private readonly IMapManager _mapManager = default!;
+        [Dependency] private readonly SolutionContainerSystem _solutionSystem = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<AbsorbentComponent, AfterInteractEvent>(OnAfterInteract);
+        SubscribeLocalEvent<TransferCancelledEvent>(OnTransferCancelled);
+        SubscribeLocalEvent<TransferCompleteEvent>(OnTransferComplete);
+    }
+
+    private void OnAfterInteract(EntityUid uid, AbsorbentComponent component, AfterInteractEvent args)
+    {
+        var user = args.User;
+        var used = args.Used;
+        var target = args.Target;
+
+        _solutionSystem.TryGetSolution(used, "absorbed", out var absorbedSolution);
+
+        var toolAvailableVolume = FixedPoint2.New(0);
+        var toolCurrentVolume = FixedPoint2.New(0);
+        var transferAmount = FixedPoint2.New(0);
+
+        if (absorbedSolution is not null)
+        {
+            toolAvailableVolume = absorbedSolution.AvailableVolume;
+            toolCurrentVolume = absorbedSolution.CurrentVolume;
+        }
```

Comment:
you should probably just exit if `absorbedSolution` is null, tbh. that makes this a little cleaner